sync.go

```diff
@@ -132,7 +132,11 @@ func (w *PipeWriter) Write(p []byte) (n int, err error) {
 			}
 
 			// chunk write to fill space
-			m, err = w.b.Write(p[n : int64(n)+gap(w.b)])
```

Comment:
Why don't we break out this inner for loop if gap(w.b) >= int64(len(p[n:]), it'll then fall through to just do a std. write, and the write should fit entirely in the buffer (we won't lose/regain the lock again).